### PR TITLE
改訂版SBP実装（On-policy版）

### PR DESCRIPTION
**Pull request by ciela**
#11 について、GitHubが障害中に修正分をpushしてしまったせいか、最後のコメントからの更新分のコミットが差分に反映されなくなってしまいました…
こちらに改めてPRを作成しましたのでレビューはこちらでお願いします :bow: 
#11 のPRはクローズしておきます。

### CODE REVIEW

File: joeynmt/model.py [hunk 1]
```diff
@@ -519,42 +525,97 @@ def adoption_model(log_prob: Tensor, tau: Tensor) -> Tensor:
...
-            src_mask = src_mask.index_select(0, adopted_indexes)
-            log_probs = log_probs[to_adopt].unsqueeze(dim=1)
-            trg = trg.index_select(0, adopted_indexes)
-            length_norms = length_norms.index_select(0, adopted_indexes)
-            distributions.append(Categorical(logits=logits.index_select(0, adopted_indexes)))
-            # adoption end
+            # if current length has reached to beam max length, continue with only greedy search
+            if l >= beam_maxlen:
+                # just get argmax
+                next_log_probs, next_ys_tokens = log_probs.max(dim=1)
```
Comment (ciela) [inline joeynmt/model.py:531]: argmaxを取るのは `log_probs` （これまでの `log_probs` の合計）でよろしかったでしょうか？
（擬似コード通りにすると `logits` をベースにするべき？とも思ったのですが一応確認です）

File: joeynmt/model.py [hunk 2]
```diff
@@ -519,42 +525,97 @@ def adoption_model(log_prob: Tensor, tau: Tensor) -> Tensor:
...
+                greedy_thresholds = thresholds[greedy]
+                greedy_log_probs = log_probs[greedy]
+                greedy_log_probs_norm = log_probs_norm[greedy]
+                greedy_length_norms = length_norms[greedy]
+                greedy_encoder_output = encoder_output[greedy]
+                greedy_src_mask = src_mask[greedy]
+                greedy_trg = trg[greedy]
+                # just get argmax
+                greedy_next_log_probs, greedy_next_ys_tokens = greedy_log_probs.max(dim=1)
+                greedy_ys_tokens = torch.cat((greedy_ys_tokens, greedy_next_ys_tokens.unsqueeze(1)), dim=1)
```
Comment (ciela) [inline joeynmt/model.py:553]: （#11 からの再掲です）
EOSシーケンスに対してはデコーダにて次のトークンもEOS固定になるため、擬似コードにおけるrolloutに入ってからのEOSに対するif分岐は実装上は特に必要はなく全て愚直にargmaxすればOKだという認識ですが問題なさそうでしょうか？

File: joeynmt/model.py (hunk 1 above)
Comment (tetsuromorimura) [reply joeynmt/model.py:531]: `log_probs`と`logits`のどちらでargmaxを取っても、返ってくる結果は同じはずななので、こちらの実装でOKです。

あるトークン列に対して、新しいトークンを追加する場合に、どのトークンが最も確率が高いかを計算していて、新しいトークンを追加する前の元のトークン列（の`log_probs`）は、比較されるトークン列間で等しいはずなので

File: joeynmt/model.py (hunk 2 above)
Comment (tetsuromorimura) [reply joeynmt/model.py:553]: ご指摘の通りで、全て愚直にargmaしてOKです！

**Review by tetsuromorimura — APPROVED** [review]
確認しました！OKそうです！
